src/commons/editor/useComments.tsx

```diff
@@ -0,0 +1,261 @@
+    (...commentsToRemove: IComment[]) => {
+      const ids = commentsToRemove.map(c => c.id);
+      const updated = omit(commentsBeingEdited, ids);
+      setCommentsBeingEdited(updated);
+      commentEditChangeEE.emit('delete', updated, ids);
+    },
+    [commentEditChangeEE, commentsBeingEdited, setCommentsBeingEdited]
+  );
+
+  const removeComment = React.useCallback(
+    (...commentsToRemove: IComment[]) => {
+      const ids = commentsToRemove.map(c => c.id);
+      setComments(omit(comments, ids));
+      // Also remove from being edited.
+      removeCommentEdit(...commentsToRemove);
+    },
+    [comments, removeCommentEdit]
+  );
+
+  const isUnsubmittedComment = React.useCallback((comment: IComment) => {
+    return comment.datetime === Infinity;
+  }, []);
+
+  const API: CommentAPI = {
+    updateComment,
+    updateCommentsBeingEdited,
+    removeComment,
+    removeCommentEdit,
+    isUnsubmittedComment
+  };
+
+  const APIRef = React.useRef(API);
+  APIRef.current = API;
+
+  // ----------------- ACTUAL OPERATIONS -----------------
+
+  const widgetManagerRef: React.MutableRefObject<ILineManager | null> = React.useRef(null);
+  React.useEffect(() => {
+    if (!reactAceRef.current) {
+      return;
+    }
+    const editor = reactAceRef.current!.editor;
+    widgetManagerRef.current = new LineWidgets(editor.session);
+    widgetManagerRef.current!.attach(editor);
+  }, [reactAceRef]);
+
+  const createCommentPrompt = React.useCallback(
+    row => {
+      const id = uuidv1();
+      const newcomment = {
+        id,
+        isCollapsed: false,
+        username: 'My user name',
+        profilePic: 'https://picsum.photos/200',
+        linenum: row,
+        text: '',
+        datetime: Infinity
+        // Not submitted yet!
+        // Will sort to the end.
+      };
+      setComments({
+        ...comments,
+        [id]: newcomment
+      });
+      setCommentsBeingEdited({
+        ...commentsBeingEdited,
+        [id]: newcomment
+      });
+    },
+    [comments, commentsBeingEdited, setCommentsBeingEdited]
+  );
+
+  contextMenuHandlers.createCommentPrompt = createCommentPrompt;
+
+  // ----------------- RENDERING -----------------
+
+  // Render comments.
+  React.useEffect(() => {
+    // React can't handle the rendering because it's going into
+    // an unmanaged component.
+    // Also, the line number changes externally, so extra fun.
+    // TODO: implement line number changes for comments.
+    // TODO: Put a minimize/maximize button.
+
+    // Re-render all comments.
+    const commentsByLine = groupBy(values(comments), c => c.linenum);
+    const commentsWidgets = map(commentsByLine, commentsOnLineUnsorted => {
+      const commentsOnLine = sortBy(commentsOnLineUnsorted, c => c.datetime);
+      const container = document.createElement('div');
+      container.style.maxWidth = '40em';
+      // container.style.backgroundColor = 'grey';
+      const widget: IWidget = {
```

Comment:
I assume this `IWidget` is an interface exposed by Ace editor for custom plugins to work with individual lines in the editor, albeit undocumented?

Reply:
It's their own internal api i reversed engineered...

Reply:
Oh my, that's some... bad SWE on their part D: